### ISSUE TITLE
Can an icon be added to the left of the interface declaration and methods for quick jumps to related implementations?

### ISSUE THREAD

**Issue by life2015** [redhat-developer/vscode-java#3524]
Can an icon be added to the left of the interface declaration and methods for quick jumps to related implementations? ( and an icon for quick jump to related Interface)

> example from IntelliJ IDEA:
![image](https://github.com/redhat-developer/vscode-java/assets/12879047/74004a0b-4c4d-4a33-81d1-d5936c998509)
<img width="439" alt="image" src="https://github.com/redhat-developer/vscode-java/assets/12879047/5cbdc8ea-23cc-4d10-9ecb-ce7529d8fb7e">

##### Environment
- Operating System: Linux
- JDK version: 17
- Visual Studio Code version: 1.87.0
- Java extension version: v1.28.1

##### Steps To Reproduce
1. Create a java interface
2. Create an implementation of this interface

##### Current Result
<img width="329" alt="image" src="https://github.com/redhat-developer/vscode-java/assets/12879047/485d1d6d-11a8-49c0-8d86-ba6ec2172250">

##### Expected Result
![image](https://github.com/redhat-developer/vscode-java/assets/12879047/c806383c-114e-49a6-8d62-f0a396f9ca5d)

##### Additional Informations
In Java development, many people are accustomed to using the icons next to Interfaces to jump to the implementations of the interface, or using the related icons inside a Class to jump to the corresponding Interface definition. The support for this feature in vscode-java can facilitate a smoother adaptation to these user habits.

**Comment by rgrunber** [comment]
@jdneo , does VS Code have API to support this ? We have "Go to declaration" which will do this if a key binding is set.

**Comment by jdneo** [comment]
> @jdneo, does VS Code have API to support this?

Yes, we already implemented some gutter decorations in the Spring Dashboard extension: https://github.com/microsoft/vscode-spring-boot-dashboard/blob/main/src/gutter/index.ts.

'Jumping to implementation' is one of the frequently mentioned feedback in our survey. Looks like it's still a problem for users to get familiar with VS Code UI when they get used to other Java IDEs.

On the other hand, before starting implement this, we need to evaluate the perf of updating the gutter when document changes. It might be heavy?

**Comment by life2015** [comment]
> > @jdneo, does VS Code have API to support this?
> 
> Yes, we already implemented some gutter decorations in the Spring Dashboard extension: https://github.com/microsoft/vscode-spring-boot-dashboard/blob/main/src/gutter/index.ts.
> 
> 'Jumping to implementation' is one of the frequently mentioned feedback in our survey. Looks like it's still a problem for users to get familiar with VS Code UI when they get used to other Java IDEs.
> 
> On the other hand, before starting implement this, we need to evaluate the perf of updating the gutter when document changes. It might be heavy?

The Java plugin currently has a feature "implementationsCodeLens", which I think has the same performance overhead as the aforementioned "Jumping to implementation", just through a different interaction method. Perhaps we could start by setting a switch to turn this feature on, to avoid potential performance issues caused by it being enabled by default.

**Comment by jdneo** [comment]
> Perhaps we could start by setting a switch to turn this feature on, to avoid potential performance issues caused by it being enabled by default.

For sure, we can add a switch for this feature. While, based on past experience, it's difficult for users to discover a feature that is default disabled. Finally, we need to solve the perf issue (if it has).

One idea from my side is first do some profiling for the feature `implementationsCodeLens` to see if the cost is acceptable and where we can improve.
